frame/merkle-mountain-range/src/lib.rs

```diff
@@ -196,13 +206,26 @@ pub mod pallet {
 	#[pallet::getter(fn mmr_leaves)]
 	pub type NumberOfLeaves<T, I = ()> = StorageValue<_, LeafIndex, ValueQuery>;
 
-	/// Hashes of the nodes in the MMR.
+	/// Leaf (data) and parent nodes (hashes) added by current block.
+	///
+	/// This is single value storage overwritten by each block. It is used as temporary
+	/// leaf storage until after block is built and offchain worker can move/copy this
+	/// leaf and its parent nodes in offchain database under `node_offchain_key()`.
```

Comment:
I. yes, offchain worker is guaranteed to run after every block; mmr offchain canonicalization is also based on same assumption (as are other pallets)

> when offchain_worker() runs for block n, are we guaranteed that the runtime didn't start producing block n+1 and overwrote NewNodes ? Is this synchronous ?

II. `offchain_worker(n)` will run before next block production, yes; but even if that wasn't the case, it would still work because offchain_worker runs with runtime storage/state fixed at block `n` (nodes have a `--blocks-pruning <N>` parameter, configuring `N` as for how many recent blocks should runtime state be held in database - the node can access exact state as it was at block `latest - N <= b <= latest`).